### PR TITLE
Proc macro type

### PR DESCRIPTION
**Pull request by P-E-P**
Add a proc_macro type to identify the location of a proc macro function in the proc macro library. This model is slightly based upon the rustc format, the main difference resides within the fact we do not use a bridge/server communication but instead a function pointer directly to the macro function.

### CODE REVIEW

File: libgrust/libproc_macro/proc_macro.h [hunk 1]
```diff
@@ -23,6 +23,69 @@
...
+  BangPayload bang;
+};
+
+struct Procmacro
+{
+  ProcmacroTag tag;
+  ProcmacroPayload payload;
+
+public:
+  Procmacro make_derive (char *trait_name, char **attribute, std::uint64_t size,
```
Comment (philberty) [inline libgrust/libproc_macro/proc_macro.h:83]: can this be const char* trait_name? where possible

File: libgrust/libproc_macro/proc_macro.cc [hunk 2]
```diff
@@ -21,3 +21,32 @@
...
 #include "proc_macro.h"
+
+namespace Procmacro {
+
+Procmacro
+Procmacro::make_derive (char *trait_name, char **attributes, std::uint64_t size,
+			CustomDeriveMacro macro)
+{
+  ProcmacroPayload payload;
+  payload.custom_derive = {trait_name, attributes, size, macro};
```
Comment (philberty) [inline libgrust/libproc_macro/proc_macro.cc:32]: I'm wondering should you map this over to a std::string in the struct so we dont have to worry about doing things like strdup

**Review by philberty — CHANGES_REQUESTED** [review]
I think if the structs were managing within the extern C API we need to be careful with the string pointers here. Iff they are all internal strings then maybe its ok but if we are getting input from else where you might want to think about the memory management here and use strdup to take copies of the strings and have cleanup functions for the structs.

I would however be super wrong about whats going on lol

File: libgrust/libproc_macro/proc_macro.cc (hunk 2 above)
Comment (P-E-P) [reply libgrust/libproc_macro/proc_macro.cc:32]: To be fair I'm not even sure we need the constructors, these structures should be embedded in the final proc_macro library and thus will be directly mapped to the file's content. The strings won't ever be created from us but rather by the backend from the trait name etc... in the ast.

File: libgrust/libproc_macro/proc_macro.cc [hunk 3]
```diff
@@ -21,3 +21,32 @@
 // <http://www.gnu.org/licenses/>.
 
 #include "proc_macro.h"
+
+namespace Procmacro {
```
Comment (CohenArthur) [inline libgrust/libproc_macro/proc_macro.cc:25]:
```suggestion
namespace ProcMacro {
```
nit, feel free to ignore

**Review by CohenArthur — COMMENTED** [review]
This looks good to me. I think we'll need to start using the abstractions to figure out how good it is to interact with them. I had the same concerns as @philberty regarding strings and memory allocations but I see what you mean by "the strings won't be created from us, but simply borrowed from the AST". In that case however, I'd like to see more `const char *` pointers if possible - since we're just borrowing, and we're not going to change these strings, it seems to me that it should be possible.

**Comment by P-E-P** [comment]
> This looks good to me. I think we'll need to start using the abstractions to figure out how good it is to interact with them. I had the same concerns as @philberty regarding strings and memory allocations but I see what you mean by "the strings won't be created from us, but simply borrowed from the AST". In that case however, I'd like to see more `const char *` pointers if possible - since we're just borrowing, and we're not going to change these strings, it seems to me that it should be possible.

Is it even worth keeping those constructors ?

**Comment by CohenArthur** [comment]
> > This looks good to me. I think we'll need to start using the abstractions to figure out how good it is to interact with them. I had the same concerns as @philberty regarding strings and memory allocations but I see what you mean by "the strings won't be created from us, but simply borrowed from the AST". In that case however, I'd like to see more `const char *` pointers if possible - since we're just borrowing, and we're not going to change these strings, it seems to me that it should be possible.
> 
> Is it even worth keeping those constructors ?

I think so, yeah :) we can always remove them later on

**Review by CohenArthur — APPROVED** [review]

**Pull request merged**